src/server/pkg/sync/sync.go

```diff
@@ -19,17 +18,16 @@ import (
 type Puller struct {
 	// errCh contains an error from the pipe goros
 	errCh chan error
-	// pipes is a set containing all pipes that are currently blocking
-	pipes map[string]bool
-	// pipesMu is a mutex to synchronize access to pipes
-	pipesMu sync.Mutex
+	// pipes is a set containing all pipes created
+	pipes []string
+	// cleaningUp is true if Puller is in the middle of a cleanUp process
+	cleaningUp bool
```

Comment:
It is but it's used in an implicitly cooperative way.
The goro that calls `CleanUp` is the only goro that writes to it, setting it to `true`.
That same goro then opens all of the pipes which unblocks the other goros which notice that we're cleaning up and exit. I'm pretty sure that this behavior is entirely deterministic.